Mathlib/Logic/Equiv/Basic.lean

```diff
@@ -0,0 +1,1900 @@
+  ⟨fun x => x, fun x => ⟨x, h x⟩, fun _ => Subtype.eq rfl, fun _ => rfl⟩
+#align equiv.subtype_univ_equiv Equiv.subtypeUnivEquiv
+
+/-- A subtype of a sigma-type is a sigma-type over a subtype. -/
+def subtypeSigmaEquiv (p : α → Type v) (q : α → Prop) : { y : Sigma p // q y.1 } ≃ Σ x :
+    Subtype q, p x.1 :=
+  ⟨fun x => ⟨⟨x.1.1, x.2⟩, x.1.2⟩, fun x => ⟨⟨x.1.1, x.2⟩, x.1.2⟩, fun _ => rfl,
+    fun _ => rfl⟩
+#align equiv.subtype_sigma_equiv Equiv.subtypeSigmaEquiv
+
+/-- A sigma type over a subtype is equivalent to the sigma set over the original type,
+if the fiber is empty outside of the subset -/
+def sigmaSubtypeEquivOfSubset (p : α → Type v) (q : α → Prop) (h : ∀ x, p x → q x) :
+    (Σ x : Subtype q, p x) ≃ Σ x : α, p x :=
+  (subtypeSigmaEquiv p q).symm.trans <| subtypeUnivEquiv fun x => h x.1 x.2
+#align equiv.sigma_subtype_equiv_of_subset Equiv.sigmaSubtypeEquivOfSubset
+
+/-- If a predicate `p : β → Prop` is true on the range of a map `f : α → β`, then
+`Σ y : {y // p y}, {x // f x = y}` is equivalent to `α`. -/
+def sigmaSubtypeFiberEquiv {α β : Type _} (f : α → β) (p : β → Prop) (h : ∀ x, p (f x)) :
+    (Σ y : Subtype p, { x : α // f x = y }) ≃ α :=
+  calc
+    _ ≃ Σy : β, { x : α // f x = y } := sigmaSubtypeEquivOfSubset _ p fun _ ⟨x, h'⟩ => h' ▸ h x
+    _ ≃ α := sigmaFiberEquiv f
+#align equiv.sigma_subtype_fiber_equiv Equiv.sigmaSubtypeFiberEquiv
+
+/-- If for each `x` we have `p x ↔ q (f x)`, then `Σ y : {y // q y}, f ⁻¹' {y}` is equivalent
+to `{x // p x}`. -/
+def sigmaSubtypeFiberEquivSubtype {α β : Type _} (f : α → β) {p : α → Prop} {q : β → Prop}
+    (h : ∀ x, p x ↔ q (f x)) : (Σ y : Subtype q, { x : α // f x = y }) ≃ Subtype p :=
+  calc
+    (Σy : Subtype q, { x : α // f x = y }) ≃ Σy :
+        Subtype q, { x : Subtype p // Subtype.mk (f x) ((h x).1 x.2) = y } := by {
+          apply sigmaCongrRight
+          intro y
+          apply Equiv.symm
+          refine' (subtypeSubtypeEquivSubtypeExists _ _).trans (subtypeEquivRight _)
+          intro x
+          exact ⟨fun ⟨hp, h'⟩ => congr_arg Subtype.val h', fun h' => ⟨(h x).2 (h'.symm ▸ y.2),
+            Subtype.eq h'⟩⟩ }
+    _ ≃ Subtype p := sigmaFiberEquiv fun x : Subtype p => (⟨f x, (h x).1 x.property⟩ : Subtype q)
+#align equiv.sigma_subtype_fiber_equiv_subtype Equiv.sigmaSubtypeFiberEquivSubtype
+
+/-- A sigma type over an `Option` is equivalent to the sigma set over the original type,
+if the fiber is empty at none. -/
+def sigmaOptionEquivOfSome (p : Option α → Type v) (h : p none → False) :
+    (Σ x : Option α, p x) ≃ Σ x : α, p (some x) :=
+  haveI h' : ∀ x, p x → x.isSome := by
+    intro x
+    cases x
+    · intro n
+      exfalso
+      exact h n
+    · intro _
+      exact rfl
+  (sigmaSubtypeEquivOfSubset _ _ h').symm.trans (sigmaCongrLeft' (optionIsSomeEquiv α))
+#align equiv.sigma_option_equiv_of_some Equiv.sigmaOptionEquivOfSome
+
+-- ericr: this definition doesn't seem nice indentation-wise; is this just the Lean4 style?
```

Comment:
I'm surprised this is the one that bothers you :)

Maybe all the equiv definitions that don't fit on one line should use the `where` syntax